Add entity_picture_template options to Template Cover, Template Light, Template Sensor, and Template Switch

bachya opened the pull request:
## Description:
__(redoing a messed up PR: https://github.com/home-assistant/home-assistant/pull/9792)__

This is a PR to provide an `entity_picture_template` configuration option to all Template platforms (except for the Template Binary Sensor, which derives its iconography from its device class). 

Like icons, the use case is to be able to change entity pictures based on Jinja templates – for example, a device tracker might show a colorized version of a person's picture when they're home and a monochromatic version when they aren't.

Additionally, this PR adds `icon_template` support to the Template Light since it wasn't present.

**Related issue (if applicable):** N/A

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** https://github.com/home-assistant/home-assistant.github.io/pull/3654

## Example entry for `configuration.yaml` (if applicable):
```yaml
cover:
  - platform: template
    covers:
      garage_door:
        friendly_name: "Garage Door"
        value_template: "{{ states('sensor.garage_door')|float > 0 }}"
        open_cover:
          service: script.open_garage_door
        close_cover:
          service: script.close_garage_door
        stop_cover:
          service: script.stop_garage_door
        entity_picture_template: >-
          {% if is_state('cover.garage_door', 'True') %}
            /local/images/garage_open.png
          {% else %}
            /local/images/garage_closed.png
          {% endif %}

light:
  - platform: template
    lights:
      theater_lights:
        friendly_name: "Theater Lights"
        level_template: "{{ sensor.theater_brightness.attributes.lux|int }}"
        value_template: "{{ sensor.theater_brightness.attributes.lux|int > 0 }}"
        turn_on:
          service: script.theater_lights_on
        turn_off:
          service: script.theater_lights_off
        set_level:
          service: script.theater_lights_level
          data_template:
            brightness: "{{ brightness }}"
        entity_picture_template: >-
          {% if is_state('light.theater_lights', 'True') %}
            /local/images/light_on.png
          {% else %}
            /local/images/light_off.png
          {% endif %}

sensor:
  - platform: template
    sensors:
      sunrise:
        value_template: "{{ states.sun.sun.attributes.next_rising }}"
      entity_picture_template: >-
          {% if is_state('sensor.sunrise', 'True') %}
            /local/images/sunrise.png
          {% else %}
            /local/images/sunset.png
          {% endif %}

switch:
  - platform: template
    switches:
      skylight:
        value_template: "{{ is_state('sensor.skylight', 'on') }}"
        turn_on:
          service: switch.turn_on
          data:
            entity_id: switch.skylight_open
        turn_off:
          service: switch.turn_on
          data:
            entity_id: switch.skylight_close
        entity_picture_template: >-
          {% if is_state('switch.skylight', 'True') %}
            /local/images/on.png
          {% else %}
            /local/images/off.png
          {% endif %}
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

bachya committed:
Re-adding cover

bachya committed:
Re-adding light

bachya committed:
Re-adding sensor

bachya committed:
Re-adding switch

bachya committed:
Re-added tests

bachya commented:
@pvizeli: FYI, I messed up my previous PR with a bad rebase; re-creating it (with your requested changes) here.

bachya commented:
Documentation complete; removing WIP status.

pvizeli commented on homeassistant/components/cover/template.py:
```diff
@@ -374,16 +392,28 @@ def async_update(self):
             try:
-                self._icon = self._icon_template.async_render()
+                setattr(self, '_{0}'.format(property_name),
+                        icon_picture_template.async_render())
```
Simplify this. Maybe with a helper function. It's very hard to read. Or do it with a MAP helper and loop it.

bachya replied:
It seems pretty simple; unclear on your suggestions – could you share an example?

bachya replied:
@pvizeli After taking a look through, I'm still unclear on what exactly needs to change. Putting this into a helper function just abstracts the complexity; it doesn't remove it. Can you assist with further info?

pvizeli commented on homeassistant/components/light/template.py:
```diff
@@ -241,3 +275,28 @@ def async_update(self):
+        if icon_picture_template is not None:
+            try:
+                setattr(self, '_{0}'.format(property_name),
+                        icon_picture_template.async_render())
```
Same

bachya committed:
Fixing missing imports in rebased test

bachya committed:
Fixing broken tests

bachya commented:
@pvizeli: I've simplified those methods as best I can; open to other ideas if you'd like further simplification.

balloob commented on homeassistant/components/cover/template.py:
```diff
@@ -237,6 +249,11 @@ def icon(self):
     @property
+    def entity_picture(self):
+        """Return the entity picture to use in the frontend, if any."""
+        return self._icon
```
This can never be the same value as the icon property 🤔

balloob commented on homeassistant/components/cover/template.py:
```diff
@@ -369,9 +386,12 @@ def async_update(self):
                 self._tilt_value = None
-        if self._icon_template is not None:
+
+        icon_template = (self._icon_template if self._icon_template else
```
Please just write normal if statements, by getting them multiline it is very difficult to read.

balloob commented on homeassistant/components/cover/template.py:
```diff
@@ -369,9 +386,12 @@ def async_update(self):
+
+        icon_template = (self._icon_template if self._icon_template else
+                         self._entity_picture_template)
+        if icon_template is not None:
```
This is wrong. The `icon_template` can now be an entity picture template too ? And then it renders to icon property? That makes no sense

balloob commented on homeassistant/components/light/template.py:
```diff
@@ -146,6 +170,16 @@ def should_poll(self):
+    @property
+    def entity_picture(self):
+        """Return the entity picture to use in the frontend, if any."""
+        return self._icon
```
This has to be fixed for all.

balloob commented on homeassistant/components/sensor/template.py:
```diff
@@ -100,13 +103,13 @@ def template_sensor_state_listener(entity, old_state, new_state):
             """Update template on startup."""
-            async_track_state_change(
-                self.hass, self._entities, template_sensor_state_listener)
+            async_track_state_change(self.hass, self._entities,
```
Unnecessary change

balloob commented on homeassistant/components/sensor/template.py:
```diff
@@ -100,13 +103,13 @@ def template_sensor_state_listener(entity, old_state, new_state):
 
-        self.hass.bus.async_listen_once(
-            EVENT_HOMEASSISTANT_START, template_sensor_startup)
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START,
```
Unnecessary change

balloob commented on homeassistant/components/sensor/template.py:
```diff
@@ -124,6 +127,11 @@ def icon(self):
     @property
+    def entity_picture(self):
+        """Return the entity_picture to use in the frontend, if any."""
+        return self._icon
```
Entity picture != icon

balloob commented on homeassistant/components/sensor/template.py:
```diff
@@ -148,9 +156,11 @@ def async_update(self):
             _LOGGER.error('Could not render template %s: %s', self._name, ex)
 
-        if self._icon_template is not None:
+        icon_template = (self._icon_template if self._icon_template else
```
No multi line if else

balloob commented on homeassistant/components/switch/template.py:
```diff
@@ -162,9 +177,11 @@ def async_update(self):
             self._state = None
 
-        if self._icon_template is not None:
+        icon_template = (self._icon_template if self._icon_template else
```
🐛

bachya commented:
@balloob: Thanks for your feedback. All of your suggestions apply to 2nd-round changes I made after @pvizeli gave me feedback that my code was too complex; my original commits addressed all of your points. I asked for ways I might simplify, but didn’t receive feedback. Happy to return to what I had?

balloob commented:
The problem with your original code was that it was confusing to follow. Pascal meant the map function, that would be a better solution.

```python
for property_name, template in (
        ('_icon', self.icon_template),
        ('_entity_picture', self.picture_template)):
    …
```

balloob commented:
This PR would have also been a lot easier if you started with 1 platform instead of trying to do all at the same time.

bachya commented:
Thanks, @balloob; that helps. I'll keep the "atomic PR" in mind for the future. In the meantime, I'll revert to where I was, simplify that logic, and we can go from there.

bachya committed:
Owner-requested changes

balloob commented on homeassistant/components/cover/template.py:
```diff
@@ -369,16 +387,26 @@ def async_update(self):
+            if template is None:
+                continue
+
+            friendly_property_name = property_name[1:].replace('_', ' ')
```
Only calculate this if you need it: inside the exception handler.

bachya committed:
Owner-requested changes

bachya committed:
Fixed exception

balloob commented:
Thanks! 🐬  🌮
